rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,151 @@
+However, applications are one piece of the UI code in a browser, and they have
+dependencies on other UI plugins. Apps can't control plugins and other things
+that Kibana loads in the browser.
+
+This RFC proposes a Screenshot Mode Service as a low-level plugin that allows
+other plugins (UI code) to make choices when the page is rendering for a screenshot.
+
+More background on how Reporting currently works, including the lifecycle of
+creating a PNG report, is here: https://github.com/elastic/kibana/issues/59396
+
+# Motivation
+
+The Reporting team wants all applications to support a customized URLs, such as
+Canvas does with its `#/export/workpad/pdf/{workpadId}` UI route. The
+customized URL is where an app can solve any rendering issue in a PDF or PNG,
+without needing extra CSS to be injected into the page.
+
+However, many low-level plugins have been added to the UI over time. These run
+on every page and an application can not turn them off. Reporting performance
+is negatively affected by this type of code. When the Reporting team analyzes
+customer logs to figure out why a job timed out, we sometimes see requests for
+the newsfeed API and telemetry API: services that aren't needed during a
+reporting job.
+
+In 7.12.0, using the customized `/export/workpad/pdf` in Canvas, the Sample
+Data Flights workpad loads 163 requests. Most of thees requests don't come from
+the app itself but from the application container code that Canvas can't turn
+off.
+
+# Detailed design
+
+The Screenshot Mode Service is an entirely new plugin that has an API method
+that returns a Boolean. The return value tells the plugin whether or not it
+should render itself to optimize for non-interactivity.
+
+The plugin is low-level as it has no dependencies of its own, so other
+low-level plugins can depend on it.
+
+## Interface
+A plugin would depend on `screenshotMode` in kibana.json. That provides
+`screenshotMode` as a plugin object. The plugin's purpose is to know when the
+page is rendered for screenshot capture, and to interact with plugins through
+an API. It allows plugins to decides what to do with the screenshot mode
+information.
+
+```
+interface IScreenshotModeServiceSetup {
+  isScreenshotMode: () => boolean;
+}
+```
+
+The plugin knows the screenshot mode from request headers: this interface is
+constructed from a class that refers to information sent via a custom
+proprietary header:
+
+```
+interface HeaderData {
+  'X-Screenshot-Mode': true
+}
+
+class ScreenshotModeServiceSetup implements IScreenshotModeServiceSetup {
+  constructor(rawData: HeaderData) {}
+  public isScreenshotMode (): boolean {}
+}
+```
+
+The Reporting headless browser that opens the page can inject custom headers
+into the request. Teams should be able to test how their app renders when
+loaded with this header. They could use a web debugging proxy, or perhaps the
+new service should support a URL parameter which triggers screenshot mode to be
+enabled, for easier testing.
```

Comment:
Should we use this param not just for testing? It would be like a real alternative to the header.

Reply:
There needs to be some detail about how the client-side of plugin is aware that the page load request had a special header that is relevant. I imagined there would be a server-side piece that acts like middleware to requests: runs on every request, keeps its own state, and is able to render the state into the DOM before the initial HTML is returned. The model would be like how the client-side UiSettings service works.

Looking at that now, there could be invalid assumptions about what a server-side plugin can do. I avoided proposing a URL parameter because my thinking was the URL query string is reserved for the app. If plugins are modifying it, it could create conflicts with the app. I could be wrong though.

Reply:
@tsullivan, is there any reason at all for it to be a header? Or could it be just a part of the URL? If it is just URL then the whole setup seems simpler. 

> If plugins are modifying it, it could create conflicts with the app.

I think `screenshotMode` plugin could pick it up in setup and put the flag into memory.
or
If `screenshotMode` is not a plugin, but part of core, then it could check and persist that query param before running any plugin, so no one would have a chance to mess it up.

Reply:
i think we should think thru the internal implementation as well, once we agree on the public interface, which seems quite simple:

```
/***
* Call this method if you are a lower level plugin without an app and you want a quick way to disable yourself when kibana is in screenshot mode. Try to avoid using this, talk to app-services team when in doubt
***/
isScreenshotMode: () => boolean
```

so how do we get that information in there, without doing some weird hacky thing that's gonna be the source of errors and is gonna work on client and server.

Reply:
@ppisljar Could you please give an example of server usage?

Reply:
yeah good point, lets not over complicate :)